bytecomp/translmod.ml

```diff
@@ -342,9 +353,82 @@ let transl_class_bindings cl_list =
        (id, transl_class ids id meths cl vf))
      cl_list)
 
+(* Compile one or more functors, merging curried functors to produce
+   multi-argument functors.  Any [@inline] attribute on a functor that is
+   merged must be consistent with any other [@inline] attribute(s) on the
+   functor(s) being merged with.  Such an attribute will be placed on the
+   resulting merged functor. *)
+
+let merge_inline_attributes attr1 attr2 loc =
+  match Lambda.merge_inline_attributes attr1 attr2 with
+  | Some attr -> attr
+  | None -> raise (Error (loc, Conflicting_inline_attributes))
+
+let merge_functors mexp coercion root_path =
+  let rec merge mexp coercion path acc inline_attribute =
+    let finished () = acc, mexp, path, coercion, inline_attribute in
+    match mexp.mod_type with
+    | Mty_alias _ -> finished ()
+    | _ ->
+      match mexp.mod_desc with
+      | Tmod_functor (param, _, _, body) ->
+        let inline_attribute' =
+          Translattribute.get_inline_attribute mexp.mod_attributes
+        in
+        let arg_coercion, res_coercion =
+          match coercion with
+          | Tcoerce_none -> None, Tcoerce_none
+          | Tcoerce_functor (arg_coercion, res_coercion) ->
+            Some arg_coercion, res_coercion
+          | _ -> fatal_error "Translmod.merge_functors: bad coercion"
+        in
+        let loc = mexp.mod_loc in
+        let path = functor_path path param in
+        let inline_attribute =
+          merge_inline_attributes inline_attribute inline_attribute' loc
```

Comment:
@chambart and @lpw25 were discussing this on email; I'm not sure any very firm conclusion was reached except that @lpw25 is adamant that the current behaviour is ok.  We think the current behaviour is conservative and could be changed to a warning in due course if we so desire.